client/packages/programs/src/JsonForms/components/QuantityDispensed.tsx

```diff
@@ -0,0 +1,238 @@
+import React, { useEffect, useState } from 'react';
+import { rankWith, ControlProps, uiTypeIs } from '@jsonforms/core';
+import { withJsonFormsControlProps, useJsonForms } from '@jsonforms/react';
+import { FormLabel, Box } from '@mui/material';
+import {
+  useDebounceCallback,
+  DateUtils,
+  PositiveNumberInput,
+  useFormatDateTime,
+  useTranslation,
+} from '@openmsupply-client/common';
+import {
+  FORM_LABEL_COLUMN_WIDTH,
+  FORM_INPUT_COLUMN_WIDTH,
+  useZodOptionsValidation,
+} from '../common';
+import { get as extractProperty } from 'lodash';
+import { z } from 'zod';
+
+const Options = z
+  .object({
+    /**
+     * Expected quantity to be used per day.
+     * If not specified it is set to one.
+     */
+    quantityPerDay: z.number().optional(),
+    /**
+     * Field name for the remaining quantity before dispensing.
+     */
+    remainingQuantityField: z.string().optional(),
+    /**
+     * Field name for the total quantity prescribed (remaining + )
+     */
+    quantityPrescribedField: z.string().optional(),
+    /**
+     * Absolute target field name where the end of supply date should be stored.
+     * End of supply is calculated from the remainingQuantity + quantityDispensed where
+     * quantityDispensed is entered in this control.
+     */
+    endOfSupplyField: z.string().optional(),
+    /**
+     * Absolute field name of a datetime value in the data. This field is used as the base datetime
+     * to calculate the datetime when the patient runs out of medicine:
+     * baseDatetime + daysDispensed.
+     */
+    baseDatetimeField: z.string(),
+
+    totalQuantityLabel: z.string().optional(),
+  })
+  .strict();
+type Options = z.infer<typeof Options>;
+
+export const quantityDispensedTester = rankWith(
+  10,
+  uiTypeIs('QuantityDispensed')
+);
+
+const getEndOfSupply = (
+  baseTime: string,
+  remainingQuantity: number,
+  quantityDispensed: number,
+  options: Options | undefined
+): Date => {
+  const totalQuantity = remainingQuantity + quantityDispensed;
+  return DateUtils.startOfDay(
+    DateUtils.addDays(
+      new Date(baseTime),
+      totalQuantity * (options?.quantityPerDay ?? 1)
+    )
+  );
+};
+
+const UIComponent = (props: ControlProps) => {
+  const { data, handleChange, label, path, uischema } = props;
+  const [localData, setLocalData] = useState<number | undefined>(data);
+  const [remainingQuantity, setRemainingQuantity] = useState(0);
+  const [baseTime, setBaseTime] = useState<string | undefined>();
+  const { errors, options } = useZodOptionsValidation(
+    Options,
+    uischema.options
+  );
+  const dateFormat = useFormatDateTime();
+  const t = useTranslation(['common', 'programs']);
```

Comment:
neither required, nor actually working in my experience. The `common` namespace is the default fallback, so does not need to be specfiied:

```suggestion
  const t = useTranslation('programs');
```